core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTopologyMonitor.java

```diff
@@ -351,22 +392,59 @@ private void savePort(DriverChannel channel) {
     }
   }
 
-  private InetSocketAddress getBroadcastRpcAddress(AdminRow row) {
+  @Nullable
+  protected InetSocketAddress getBroadcastRpcAddress(AdminRow row) {
+    // in system.peers_v2 (Cassandra >= 4.0)
     InetAddress nativeAddress = row.getInetAddress("native_address");
     if (nativeAddress == null) {
-      // Cassandra < 4
+      // in system.peers or system.local
       InetAddress rpcAddress = row.getInetAddress("rpc_address");
       if (rpcAddress == null) {
-        // This could only happen if system.peers is corrupted, but handle gracefully
+        // This could only happen if system tables are corrupted, but handle gracefully
         return null;
       }
       return new InetSocketAddress(rpcAddress, port);
     } else {
-      Integer rowPort = row.getInteger("native_port");
+      // system.local for Cassandra >= 4.0
+      Integer rowPort = row.getInteger("rpc_port");
       if (rowPort == null || rowPort == 0) {
-        rowPort = port;
+        // system.peers_v2
+        rowPort = row.getInteger("native_port");
+        if (rowPort == null || rowPort == 0) {
+          rowPort = port;
+        }
       }
       return new InetSocketAddress(nativeAddress, rowPort);
     }
   }
+
+  /**
+   * Returns {@code true} if the given peer row is valid, and {@code false otherwise}.
+   *
+   * <p>This method must ensure that at least the row contains enough information to extract the
+   * node's broadcast RPC address; otherwise {@link #nodeInfoBuilder(AdminRow, InetSocketAddress)}
+   * may throw an exception.
+   */
+  protected boolean isPeerValid(AdminRow peerRow) {
+    boolean valid =
+        isSchemaV2
+            ? peerRow.isNotEmpty("native_address") && peerRow.isNotEmpty("native_port")
+            : peerRow.isNotEmpty("rpc_address");
```

Comment:
I think you meant `rpc_port`? The problem is that it is not present before C* 4.

Reply:
yes, `rpc_port`. So what will happen if `rpc_port` in C* > 4 will be `null`? Isn't there a risk that we will get null pointer somewhere if `rpc_port` will be null (but we didn't check that)?

Reply:
If `rpc_port` is null the port will be the default port, see [here](https://github.com/datastax/java-driver/blob/f1fe61726296dba41e5106bbb90f3839167ae352/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTopologyMonitor.java#L414). This won't make the driver error out (but it could end up being unable to connect to that node).
The problem if we start checking specific Cassandra versions is that we would be opening a can of worms: e.g. we would have to check if the node is running Cassandra 4+ but _not_ DSE 6.7 (which also reports Cassandra 4 but does not have that column).